mpu6050.c

```diff
@@ -0,0 +1,123 @@
+#include "mpu6050.h"
+#include <stdio.h>
+
+#include "hardware/i2c.h"
+#include "pico/stdlib.h"
+
+#define MPU6050_I2C_DEFAULT_ADDR   0x68
+
+static int mpu6050_wakeup(mpu6050_handle_t *handle);
+static int mpu6050_config_gyro(mpu6050_handle_t *handle, uint8_t gyro_fs);
+static int mpu6050_config_accel(mpu6050_handle_t *handle, uint8_t accel_fs);
+
+
+int mpu6050_init(mpu6050_handle_t *handle, const mpu6050_config_t *config)
+{
+	if (handle->address == 0){
+		handle->address = MPU6050_I2C_DEFAULT_ADDR;
+	}
+
+	i2c_init(handle->i2c, config->baudrate);
+	gpio_set_function(config->sda, GPIO_FUNC_I2C);
+	gpio_set_function(config->scl, GPIO_FUNC_I2C);
+	gpio_pull_up(config->sda);
+	gpio_pull_up(config->scl);
+
+
+	int ret = mpu6050_wakeup(handle);
+	if (ret != 2) {
+		printf("Error: Failed to wake up. Expected 2 bytes, got %d.\n", ret);
+		return -1;
+	}
+
+	ret = mpu6050_config_gyro(handle, config->gyro_fs);
+	if (ret != 2) {
+		printf("Error: Failed to configure gyro. Expected 2 bytes, got %d.\n", ret);
+		return -1;
+	}
+
+
+	ret = mpu6050_config_accel(handle, config->accel_fs);
+	if (ret != 2) {
+		printf("Error: Failed to configure accel. Expected 2 bytes, got %d.\n", ret);
+		return -1;
+	}
+
+	return 0;
+}
+
+static int mpu6050_wakeup(mpu6050_handle_t *handle){
+	uint8_t buffer[2];
+	buffer[0] = MPU6050_PWR_MGMT_1;
+	buffer[1] = 0x00;  // clear sleeping bit
+	int ret	  = i2c_write_blocking(handle->i2c, handle->address, buffer, 2, false);
+
+	return ret;
+}
+
+static int mpu6050_config_gyro(mpu6050_handle_t *handle, uint8_t gyro_fs){
+	uint8_t gyro_config_buf[2];
+	gyro_config_buf[0] = MPU6050_GYRO_CONFIG;
+	gyro_config_buf[1] = gyro_fs << 3; /* uses bits 4,3 */
```

Comment:
Not saying you have to do this here, but just as an aside, you can define a bit-mask for a register sub-field, and use [`__builtin_ctzg()`](https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html#index-_005f_005fbuiltin_005fctzg) to count the number of trailing zero bits in an integer, giving you the implicitly encoded bit-shift of a field!